ecosystem/sep-0010.md

```diff
@@ -120,19 +125,28 @@ To validate the challenge transaction the following steps are performed by the s
-* verify that transaction envelope has a correct signature by server's signing key;
-* verify that transaction envelope has a correct signature by the operation's source account;
+* verify that transaction contains a single [Manage Data](https://www.stellar.org/developers/guides/concepts/list-of-operations.html#manage-data) operation and its source account is not null;
 * verify that transaction sequenceNumber is equal to zero;
+* verify that transaction envelope has a correct signature by server's signing key;
+* verify that client signatures on the transactions are signers of the operations source account;
+* verify that client signatures are correct;
+* verify that client signatures provide weight that meets the required threshold(s);
 * use operations's source account to determine the authenticating client and perform any additional service-specific validations.
 
+When determining the weight of the signers the server's signer should be excluded, even if it is a signer on the account. In most cases the server should verify that the signers meet all thresholds of the account by checking that the sum of the weights is equal or greater to the low, mid and high thresholds, and fail to issue a JWT if all thresholds are not met. Some servers may have use cases to issue JWTs for less than all thresholds.
+
 Upon successful validation service responds with a session JWT, containing the following claims:
 
 * `iss` (the principal that issued a token, [RFC7519, Section 4.1.1](https://tools.ietf.org/html/rfc7519#section-4.1.1)) — a [Uniform Resource Identifier (URI)] for the issuer (`https://example.com` or `https://example.com/G...`)
 * `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account (`G...`)
 * `iat` (the time at which the JWT was issued [RFC7519, Section 4.1.6](https://tools.ietf.org/html/rfc7519#section-4.1.6)) — current timestamp (`1530644093`)
 * `exp` (the expiration time on or after which the JWT must not be accepted for processing, [RFC7519, Section 4.1.4](https://tools.ietf.org/html/rfc7519#section-4.1.4)) — a server can pick its own expiration period for the token, however 24 hours is recommended (`1530730493`)
 * `jti` (the unique identifier for the JWT, [RFC7519, Section 4.1.7](https://tools.ietf.org/html/rfc7519#section-4.1.7)) — hex-encoded hash of the challenge transaction (`f0e754c6ab988eb5fb2811c2cacc4d3d2aa4334763b8df7285ef341921e2530a`)
+* `thr`
+  * `low` — low threshold of the account
+  * `med` — medium threshold of the account
+  * `high` — high threshold of the account
+* `wht` — weight of the signers who signed the challenge transaction, excluding the weight of the server signer if it is also a signer of the account
```

Comment:
Maybe just phrase it as "weight of the client signer(s)? That seems more clear than talking about excluding the server signature.

Reply:
My concern here is an edge case where SEP-10 is being used by a service that offers signing. It's feasible the servers signing key will be a signer. It's also feasible someone will implement this calculation by iterating over the signers of the account and adding their weights if the key is included as a signer of the challenge transactions. It's subtle that in this situation the server is helping the user authenticate themselves. Maybe I could take this statement out of this section though given that I mention it in the paragraph added above.

Reply:
e1e07e7